File: plugins/extract-python-kafka/extract.py
"""Python Kafka client calls in, generic message channels and flows out."""

from __future__ import annotations

import json
import os
from collections import defaultdict
from typing import Any, Dict, List

import catalog
import kafka
from names import slug, title
from options import Options
from protocol import Builder, File, Input
from source import Project

MESSAGE = "message"


def extract(input_: Input, opts: Options, b: Builder, cwd: str = "") -> None:
    cwd = cwd or os.getcwd()
    root = os.path.abspath(os.path.join(cwd, input_.root))

    def rel(path: str) -> str:
        return os.path.relpath(path, cwd).replace(os.sep, "/")

    source = os.path.normpath(os.path.join(root, opts.source or "."))
    context = opts.context or os.path.basename(root)
    service = opts.service or os.path.basename(root)
    svc_id = context + "." + service
    project = Project(root, source, rel)
    for path, message in project.broken:
        b.warn(path, "cannot be parsed, so nothing in it is read: %s" % message)

    found = kafka.scan(project, opts.settings)
    if not found.publishes and not found.subscriptions:
        b.warn(svc_id, "no supported Python Kafka producer or consumer calls under %s" % rel(source))
    for where, expr in sorted(set(found.dynamic_topics)):
        b.warn(where, "Kafka topic is dynamic%s, so no channel is invented" % ((": `%s`" % expr) if expr else ""))
    clients = [item.client for item in found.publishes] + [item.client for item in found.subscriptions]
    for source_line in sorted({client.source for client in clients if "brokers" not in client.config}):
        b.warn(source_line or svc_id, "Kafka broker endpoints are resolved at runtime, so none are claimed in the catalog")

    sends = defaultdict(list)
    receives = defaultdict(list)
    for item in found.publishes:
        sends[item.topic].append(item)
    for item in found.subscriptions:
        receives[item.topic].append(item)
    for topic in sorted(set(sends) - set(receives)):
        b.warn(svc_id, "publishes Kafka topic `%s`, but no consumer for it is proven in this source tree" % topic)

    channels = [channel(topic, sends[topic], receives[topic]) for topic in sorted(set(sends) | set(receives))]
    flows = []
    for topic in sorted(sends):
        for index, item in enumerate(sends[topic], 1):
            flows.append(publish_flow(context, service, svc_id, item, index))
    for topic in sorted(receives):
        received_message = sends[topic][0].message if len({item.message for item in sends[topic]}) == 1 else "message"
        for index, item in enumerate(receives[topic], 1):
            flows.append(receive_flow(context, service, svc_id, item, received_message, index))
    flows.sort(key=lambda flow: flow["slug"])

    fragment: Dict[str, Any] = {
        "contexts": [
            {
                "id": context,
                "slug": context,
                "name": "",
                "summary": "",
                "services": [
                    {
                        "id": svc_id,
                        "slug": service,
                        "name": "",
                        "repo": "",
                        "path": "",
                        "readme": "",
                        "provides": [],
                        "consumes": [],
                        "aggregates": [],
                        "channels": channels,
                    }
                ],
            }
        ],
        "defs": {},
        "flows": flows,
        "adrs": [],
    }
    b.files.append(File(name=opts.out or "kafka.json", contents=json.dumps(fragment, indent=2, ensure_ascii=False) + "\n"))


def channel(topic: str, sends: List[kafka.Publish], receives: List[kafka.Subscription]) -> Dict[str, Any]:
    messages = []
    seen = set()
    for item in sends:
        key = (item.message, "send")
        if key not in seen:
            messages.append(catalog.message(item.message, title(item.message), message_doc(item), "send", item.encoding))
            seen.add(key)
    for item in receives:
        # Kafka subscriptions dispatch records, and source often does not prove
        # their schema. Match a single proven producer name; otherwise stay
        # honest and call the record a message.
        name = sends[0].message if len({sent.message for sent in sends}) == 1 else "message"
        key = (name, "receive")
        if key not in seen:
            messages.append(catalog.message(name, title(name), consumer_doc(item), "receive", item.client.encoding))
            seen.add(key)
    clients = []
    for item in list(sends) + list(receives):
        note = kafka.config_note(item.client)
        if note not in clients:
            clients.append(note)
    doc = "Kafka message stream. " + ". ".join(clients) + ". Broker-side partitions, replication and retention are not declared by client code."
    source = (sends[0].line if sends else receives[0].line) if sends or receives else ""
    return catalog.channel(topic, MESSAGE, "Kafka · " + topic, doc, messages, source)


def message_doc(item: kafka.Publish) -> str:
    bits = ["Published with %s" % item.client.library]
    if item.message_expression:
        bits.append("payload `%s`" % item.message_expression)
    if item.key:
        bits.append("key `%s`" % item.key)
    if item.headers:
        bits.append("headers `%s`" % item.headers)
    return "; ".join(bits) + "."


def consumer_doc(item: kafka.Subscription) -> str:
    group = item.client.config.get("consumer group")
    suffix = "; consumer group `%s`" % group if group else ""
    return "Subscribed with %s%s." % (item.client.library, suffix)


def handoff(topic: str, message: str, direction: str) -> Dict[str, str]:
    return {
        "kind": "message",
        "transport": "kafka",
        "channel": topic,
        "message": message,
        "direction": direction,
    }


def publish_flow(context: str, service: str, svc_id: str, item: kafka.Publish, index: int) -> Dict[str, Any]:
    flow_slug = slug("%s-kafka-publish-%s-%s-%d" % (service, item.topic, item.message, index))
    broker = "kafka-" + slug(item.topic)
    return catalog.flow(
        "flow." + flow_slug,
        flow_slug,
        "Publish %s" % title(item.message),
        "Publishes `%s` to Kafka topic `%s`." % (item.message, item.topic),
        item.module.rel,
        context,
        [
            catalog.participant(svc_id, "service", context),
            catalog.participant(broker, "broker", None, "Kafka · " + item.topic),
        ],
        [
            catalog.step(
                "publish",
                svc_id,
                broker,
                "call",
                "publish " + item.message,
                catalog.DECLARED,
                note=kafka.config_note(item.client),
                line=item.line,
                handoff=handoff(item.topic, item.message, "send"),
            )
        ],
        trigger={"kind": "unproven", "label": "source caller not proven", "confidence": "low"},
        entrypoint=item.entrypoint,
    )


def receive_flow(context: str, service: str, svc_id: str, item: kafka.Subscription, message: str, index: int) -> Dict[str, Any]:
    flow_slug = slug("%s-kafka-consume-%s-%d" % (service, item.topic, index))
    broker = "kafka-" + slug(item.topic)
    return catalog.flow(
        "flow." + flow_slug,
        flow_slug,
        "Consume from %s" % item.topic,
        "Consumes Kafka messages from `%s`." % item.topic,
        item.module.rel,
        context,
        [
            catalog.participant(broker, "broker", None, "Kafka · " + item.topic),
            catalog.participant(svc_id, "service", context),
        ],
        [
            catalog.step(
                "receive",
                broker,
                svc_id,
                "call",
                "receive message",
                catalog.DECLARED,
                note=consumer_doc(item),
                line=item.line,
                handoff=handoff(item.topic, message, "receive"),
            )
        ],
        trigger={"kind": "message", "label": "Kafka · " + item.topic, "confidence": "high"},
        entrypoint=item.entrypoint,
    )

File: plugins/extract-go-eventgrid/extract.go
package extractgoeventgrid

import (
	"encoding/json"
	"sort"
	"strings"

	"github.com/shortlink-org/portolan/catalog"
	"github.com/shortlink-org/portolan/internal/goscan"
	"github.com/shortlink-org/portolan/plugin"
)

func extract(in plugin.Input, opts Options) (plugin.Response, error) {
	b := &plugin.Builder{}
	tree, err := goscan.Read(in.Root)
	if err != nil {
		return plugin.Response{}, err
	}
	s := newScanner(tree)
	sites := s.sites()
	if len(sites) == 0 {
		b.Warn(in.Root, "no Azure Event Grid publisher call was found")
	}
	channels := s.catalog(sites, b)
	serviceID := opts.Context + "." + opts.Service
	fragment := catalog.Catalog{
		Contexts: []catalog.BoundedContext{{
			ID: opts.Context, Slug: opts.Context,
			Services: []catalog.Service{{
				ID: serviceID, Slug: opts.Service,
				Provides: []catalog.RpcService{}, Consumes: []catalog.RpcCall{}, Aggregates: []catalog.Aggregate{}, Channels: channels,
			}},
		}},
		Defs: map[string]catalog.TypeDef{}, Flows: []catalog.Flow{}, Adrs: []catalog.Adr{},
	}
	encoded, err := json.MarshalIndent(fragment, "", "  ")
	if err != nil {
		return plugin.Response{}, err
	}
	b.File(goscan.FirstNonEmpty(opts.Out, "eventgrid.json"), string(encoded)+"\n")
	return b.Response(), nil
}

type channelState struct {
	config   clientConfig
	sources  []string
	notes    map[string]bool
	messages map[string]catalog.ChannelMessage
}

func (s *scanner) catalog(sites []site, b *plugin.Builder) []catalog.Channel {
	states := map[string]*channelState{}
	for _, found := range sites {
		configs := s.clientConfigs(found.client, found.fn)
		if len(configs) == 0 {
			b.Warn(found.at.String(), "topic of "+found.method+" could not be resolved from the Event Grid client constructor")
			continue
		}
		messages := s.eventTypes(found.events, found.fn)
		for _, config := range configs {
			state := states[config.address]
			if state == nil {
				state = &channelState{config: config, notes: map[string]bool{}, messages: map[string]catalog.ChannelMessage{}}
				states[config.address] = state
			}
			state.sources = append(state.sources, found.at.String())
			state.notes[publishNote(found)] = true
			for _, name := range messages {
				state.messages[name] = catalog.ChannelMessage{Name: name, Title: name, Direction: catalog.ChannelSend}
			}
		}
	}

	addresses := make([]string, 0, len(states))
	for address := range states {
		addresses = append(addresses, address)
	}
	sort.Strings(addresses)
	out := make([]catalog.Channel, 0, len(addresses))
	for _, address := range addresses {
		state := states[address]
		sort.Strings(state.sources)
		notes := sortedKeys(state.notes)
		messageNames := make([]string, 0, len(state.messages))
		for name := range state.messages {
			messageNames = append(messageNames, name)
		}
		sort.Strings(messageNames)
		messages := make([]catalog.ChannelMessage, 0, len(messageNames))
		for _, name := range messageNames {
			messages = append(messages, state.messages[name])
		}
		out = append(out, catalog.Channel{
			Address: address, Kind: catalog.ChannelKindEvent, Title: state.config.title,
			Doc:      "Published through " + state.config.packageName + ". " + strings.Join(notes, " "),
			Messages: messages, Source: state.sources[0],
		})
	}
	return out
}

func publishNote(found site) string {
	verb := "Published"
	if found.batch {
		verb = "Published in batches"
	}
	return verb + " as " + found.schema + " by `" + goscan.LastSegment(found.fn.Receiver) + dot(found.fn.Receiver) + found.fn.Name + "`."
}

func sortedKeys(values map[string]bool) []string {
	out := make([]string, 0, len(values))
	for value := range values {
		out = append(out, value)
	}
	sort.Strings(out)
	return out
}

func dot(receiver string) string {
	if receiver == "" {
		return ""
	}
	return "."
}
